design/20210324.certificate-request-approve-deny.md

```diff
@@ -0,0 +1,191 @@
+authors:
+  - "@joshvanl"
+editor: "@joshvanl"
+creation-date: 2021-02-03
+last-updated: 2021-02-03
+status: implementable
+---
+
+# Approved and Denied Conditions
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [API Behaviour](#api-behaviour)
+- [RBAC](#rbac)
+  * [Scope](#scope)
+  * [API Group](#api-group)
+  * [Resource](#resource)
+  * [Verbs](#verbs)
+  * [Resource Names](#resource-names)
+- [cert-manager Approver](#cert-manager-approver)
+- [Certificates Controller Changes](#certificates-controller-changes)
+  * [Issuing Controller](#issuing-controller)
+<!-- /toc -->
+
+## Summary
+
+The CertificateRequest resource has [Approved and Denied
+conditions](https://github.com/jetstack/cert-manager/blob/7204284063702358164713379b07b73642ea9bec/pkg/apis/certmanager/v1/types_certificaterequest.go#L194).
+These conditions are based upon the [certificates.k8s.io
+CertificateSigningRequest](https://github.com/kubernetes/api/blob/3a2d6b5bb7f84daa2125131865eb81b151448d5b/certificates/v1/types.go#L198)
+conditions of the same name. The purpose of these conditions is so that users,
+or "approvers", are able to mark a request as either Approved or Denied. This
+condition gates an Issuer from signing the request. An Issuer waits for a
+CertificateRequest to have an Approved condition before signing. A
+CertificateRequest with a Denied condition will never be signed.
+
+
+## API Behaviour
+
+The Approved and Denied conditions are two distinct condition types on the
+CertificateRequest. These conditions must always have the status of True, and
+are mutually exclusive (i.e. a CertificateRequest cannot have an Approved and
+Denied condition simultaneously). This behaviour is enforced in the cert-manager
+Webhook component.
+
+An "approver" is an entity that is responsible for setting the Approved/Denied
+conditions. It is up to the approver's implementation as to what
+CertificateRequests are managed by that approver.
+
+The Reason field of the Approved/Denied condition should be set to *who* set the
+condition. Who can be interpreted however makes sense to the approver
+implementation. For example, it may include the API group of an approving policy
+controller, or the client agent of a manual request.
+
+The Message field of the Approved/Denied condition should be set to *why* the
+condition is set. Again, why can be interpreted however makes sense to the
+implementation of the approver. For example, the name of the resource that
+approves this request, the violations which caused the request to be denied, or
+the team to who manually approved the request.
+
+
+## RBAC
+
+Setting the Approved or Denied conditions are restricted by the approver having
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames: ["<issuer-kind>.<issuer-group>/<issuer-name or * for all issuer
```

Comment:
This is missing the namespace here (for issuer kinds that are namespaced). Also I think it should be plural `issuer-resource-name` instead of `issuer-kind` (e.g. `issuers` instead of `Issuer`). So `<issuer-resource-name>.<issuer-group>/<issuer-namespace>.<issuer-name>`.

Should we also consider using the term `signer` instead of `issuer` here? Then call out that Issuer/ClusterIssuer are types of *signer*? 🤷‍♂️

Reply:
Thanks for catching that.

Yep, the signer terminology makes sense to me.

Reply:
suggestion: can we document the difference between the signer and the issuer as part of this document?